### PR TITLE
CBMC proof for MQTTAgent_ResumeSession API

### PR DESCRIPTION
**Pull request by leegeth**
CBMC proof for MQTTAgent_ResumeSession API

### CODE REVIEW

File: test/cbmc/sources/mqtt_agent_cbmc_state.c [hunk 1]
```diff
@@ -143,3 +149,36 @@ bool isAgentSendCommandFunctionStatus( MQTTStatus_t mqttStatus )
...
+    __CPROVER_assert( pContext != NULL, "MQTT Agent Context is not NULL" );
+
+    for( i = 0; i < MQTT_AGENT_MAX_OUTSTANDING_ACKS; i++ )
+    {
+        /* Limit the packet Ids so that the range of packet ids retrieved through
+         * MQTT_PublishToResend can be limited as well. */
+        __CPROVER_assume( packetId < MAX_PACKET_ID );
+        pContext->pPendingAcks[ i ].packetId = packetId;
+
+        /*Add a publish command. */
```
Comment (yourslab) [inline test/cbmc/sources/mqtt_agent_cbmc_state.c:169]: I think this file needs to be uncrustified

Comment (leegeth) [reply test/cbmc/sources/mqtt_agent_cbmc_state.c:172]:
```suggestion
        /* Add a publish command. */
```

Comment (leegeth) [reply test/cbmc/sources/mqtt_agent_cbmc_state.c:172]: Uncrustify didn't catch this issue. Addressing it as a suggestion.

File: test/cbmc/stubs/core_mqtt_stubs.c [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        {
+            pPublishInfo = malloc( sizeof( MQTTPublishInfo_t ) );
+            __CPROVER_assume( pPublishInfo != NULL );
+
+            __CPROVER_assume( topicNameLength < 32U );
+            pPublishInfo->pTopicName = malloc( topicNameLength );
+            __CPROVER_assume( pPublishInfo->pTopicName != NULL );
+            pPublishInfo->topicNameLength = topicNameLength;
+
+            __CPROVER_assume( payloadLength < 64U );
```
Comment (yourslab) [inline test/cbmc/stubs/core_mqtt_stubs.c:133]: Should we also have defines for these magic numbers in the `Makefile`?

File: test/cbmc/sources/mqtt_agent_cbmc_state.c [hunk 3]
```diff
@@ -143,3 +149,39 @@ bool isAgentSendCommandFunctionStatus( MQTTStatus_t mqttStatus )
...
+
+    for( i = 0; i < MQTT_AGENT_MAX_OUTSTANDING_ACKS; i++ )
+    {
+        #ifdef MAX_PACKET_ID
+
+            /* Limit the packet Ids so that the range of packet ids retrieved through
+             * MQTT_PublishToResend can be limited as well. */
+            __CPROVER_assume( packetId < MAX_PACKET_ID );
+        #endif
+        pContext->pPendingAcks[ i ].packetId = packetId;
```
Comment (yourslab) [inline test/cbmc/sources/mqtt_agent_cbmc_state.c:170]: This is a cool approach

File: test/cbmc/stubs/core_mqtt_stubs.c [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    MQTTPublishInfo_t * pPublishInfo;
+    uint16_t topicNameLength;
+    size_t payloadLength;
+    static bool terminate = false;
+
+    __CPROVER_assert( pContext != NULL,
+                      "MQTT Context is not NULL." );
+
+    /* Only one packet per MQTT_ProcessLoop is received for the proof and it will be enough
+     * to prove the memory safety. The second invocation of MQTT_ProcessLoopreturns without
```
Comment (yourslab) [inline test/cbmc/stubs/core_mqtt_stubs.c:101]:
```suggestion
     * to prove the memory safety. The second invocation of MQTT_ProcessLoop returns without
```

File: test/cbmc/stubs/core_mqtt_stubs.c [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+            __CPROVER_assume( topicNameLength < 32U );
+            pPublishInfo->pTopicName = malloc( topicNameLength );
+            __CPROVER_assume( pPublishInfo->pTopicName != NULL );
+            pPublishInfo->topicNameLength = topicNameLength;
+
+            __CPROVER_assume( payloadLength < 64U );
+            pPublishInfo->pPayload = malloc( payloadLength );
+            __CPROVER_assume( pPublishInfo->pPayload != NULL );
+            pPublishInfo->payloadLength = payloadLength;
```
Comment (yourslab) [inline test/cbmc/stubs/core_mqtt_stubs.c:136]: Is it not also possible in MQTT to have a NULL payload with 0 length?

**Review by yourslab — COMMENTED** [review]
Looks good. I just had a few minor comments.

File: test/cbmc/sources/mqtt_agent_cbmc_state.c (hunk 3 above)
Comment (leegeth) [reply test/cbmc/sources/mqtt_agent_cbmc_state.c:170]: Thanks

**Review by yourslab — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Pull request merged**